graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java

```diff
@@ -137,33 +143,63 @@ public String getFingerprint() {
      * @return the list of matching streams
      */
     public List<Stream> match(Message message) {
```

Comment:
This is a bit unfortunate because right now this collection is at least copied twice for each message (one time in this method, another time in `Message#setStreams()`).